### PR TITLE
Add `ClientBuilder` and support for connecting to local spice runtime

### LINKED ISSUE #36: Add support for local Spice OSS runtime to spice-rs

### PR DESCRIPTION
**Pull request by ewgenius**
closes #36 

Added `spiceai::ClientBuilder` as a default way for creating sdk client.

Connect o to local spice runtime:

```rust
use spiceai::ClientBuilder;

#[tokio::main]
async fn main() {
  let mut client = ClientBuilder::new()
    .build()
    .await
    .unwrap();
}
```

Connect to spice.ai cloud:

```rust
use spiceai::ClientBuilder;

#[tokio::main]
async fn main() {
  let mut client = ClientBuilder::new()
    .with_api_key("API_KEY")
    .with_spiceai_cloud()
    .build()
    .await
    .unwrap();
}
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -30,11 +54,17 @@ async fn main() {
...
 ```rust
-use spiceai::Client;
+use spiceai::ClientBuilder;
 
 #[tokio::main]
 async fn main() {
-  let mut client = Client::new("API_KEY").await.unwrap();
+  let mut client = ClientBuilder::new()
+    .with_api_key("API_KEY")
+    .with_spiceai_cloud()
```
Comment (sgrebnov) [inline README.md]: Maybe just `.with_spiceai_cloud(API_KEY)`?

**Review by sgrebnov — APPROVED** [review]

File: README.md (hunk 1 above)
Comment (ewgenius) [reply README.md]: I think `api_key` should be separate - we might have it for local runtime as well in the future

File: src/client.rs [hunk 2]
```diff
@@ -95,3 +90,95 @@ impl SpiceClient {
...
+    pub fn new() -> Self {
+        Self {
+            api_key: None,
+            firecache_url: None,
+            flight_url: None,
+        }
+    }
+
+    /// Configures the `SpiceClient` to use the given API key.
+    pub fn with_api_key(mut self, api_key: &str) -> Self {
```
Comment (phillipleblanc) [inline src/client.rs]:
```suggestion
    pub fn api_key(mut self, api_key: &str) -> Self {
```

The [builder pattern](https://rust-unofficial.github.io/patterns/patterns/creational/builder.html) doesn't use `with_` prefixes for setting options.

File: src/client.rs [hunk 3]
```diff
@@ -95,3 +90,95 @@ impl SpiceClient {
...
+        self.api_key = Some(api_key.to_string());
+        self
+    }
+
+    /// Configures the `SpiceClient` to use the given Spice Firecache endpoint.
+    pub fn with_firecache_url(mut self, firecache_url: &str) -> Self {
```
Comment (phillipleblanc) [inline src/client.rs]:
```suggestion
    pub fn firecache_url(mut self, firecache_url: &str) -> Self {
```

File: src/client.rs [hunk 4]
```diff
@@ -95,3 +90,95 @@ impl SpiceClient {
...
+        self.firecache_url = Some(firecache_url.to_string());
+        self
+    }
+
+    /// Configures the `SpiceClient` to use the given Spice Flight endpoint.
+    pub fn with_flight_url(mut self, flight_url: &str) -> Self {
```
Comment (phillipleblanc) [inline src/client.rs]:
```suggestion
    pub fn flight_url(mut self, flight_url: &str) -> Self {
```

File: src/client.rs [hunk 5]
```diff
@@ -95,3 +90,95 @@ impl SpiceClient {
...
+        self.flight_url = Some(flight_url.to_string());
+        self
+    }
+
+    /// Configures the `SpiceClient` to use default Spice.ai Cloud endpoints.
+    /// Equals to calling `.with_firecache_url("https://firecache.spiceai.io")` and `.with_flight_url("https://flight.spiceai.io")`.
+    pub fn with_spiceai_cloud(mut self) -> Self {
```
Comment (phillipleblanc) [inline src/client.rs]:
```suggestion
    pub fn use_spiceai_cloud(mut self) -> Self {
```

File: src/client.rs [hunk 6]
```diff
@@ -95,3 +90,95 @@ impl SpiceClient {
...
+    }
+
+    /// Configures the `SpiceClient` to use the given Spice Flight endpoint.
+    pub fn with_flight_url(mut self, flight_url: &str) -> Self {
+        self.flight_url = Some(flight_url.to_string());
+        self
+    }
+
+    /// Configures the `SpiceClient` to use default Spice.ai Cloud endpoints.
+    /// Equals to calling `.with_firecache_url("https://firecache.spiceai.io")` and `.with_flight_url("https://flight.spiceai.io")`.
```
Comment (phillipleblanc) [inline src/client.rs]:
```suggestion
    /// Equivalent to calling `.with_firecache_url("https://firecache.spiceai.io")` and `.with_flight_url("https://flight.spiceai.io")`.
```

File: src/flight.rs [hunk 7]
```diff
@@ -115,7 +115,10 @@ impl SqlFlightClient {
         &mut self,
         query: &str,
     ) -> std::result::Result<FlightRecordBatchStream, Box<dyn Error>> {
-        self.authenticate().await?;
+        let api_key = self.api_key.clone();
+        if let Some(api_key) = api_key {
```
Comment (phillipleblanc) [inline src/flight.rs:119]:
```suggestion
        if let Some(api_key) = api_key.as_ref() {
```

Can avoid a clone here

File: src/client.rs [hunk 8]
```diff
@@ -55,12 +49,13 @@ impl SpiceClient {
     ///     let mut client = Client::new("API_KEY").await.unwrap();
     /// }
     /// ```
+    #[deprecated(note = "Use spiceai::ClientBuilder instead")]
```
Comment (phillipleblanc) [inline src/client.rs]: It's also idiomatic to add a `builder()` method on `SpiceClient` which returns a new builder. Then you can construct it like:

```rust
let client = SpiceClient::builder().flight_url("http://localhost:50051").build().await?;

File: src/flight.rs (hunk 7 above)
Comment (ewgenius) [reply src/flight.rs:119]: compiler complains:

CleanShot 2024-04-26 at 18 14 17@2x

File: src/client.rs (hunk 8 above)
Comment (ewgenius) [reply src/client.rs]: added, thanks

**Pull request merged**